daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala

```diff
@@ -16,6 +16,49 @@ import scala.annotation.tailrec
 
 private[validation] object Typing {
 
+  def typeOf(env: Env, exp: Expr): Type = { // testing entry point
+    // this implementation is stack-safe
+    runWork(env.typeOfExpr(exp))
+  }
+
+  // stack-safety achieved via a Work trampoline.
+  private sealed abstract class Work[A]
+  private object Work {
+    final case class Ret[A](v: A) extends Work[A]
+    final case class Delay[A](thunk: () => Work[A]) extends Work[A]
+    final case class Bind[A, X](work: Work[X], k: X => Work[A]) extends Work[A]
+  }
+
+  import Work.{Ret, Delay, Bind}
+
+  private def sequenceWork[A, T](works: List[Work[T]])(k: List[T] => Work[A]): Work[A] = {
```

Comment:
`sequenceWork` takes `List` because it processes the elements one by one (necessary because of the trampoline)
Yes we could work on an `ImmArray` + _index_, but it seems an unnecessary complication.